### ISSUE TITLE
New InstantAction: Trigger

### ISSUE THREAD

**Issue by MaximilianJohannesObpacher** [VDA5050/VDA5050#61]
In our opinion, the instant action waitForTrigger is defined to thight. In practise the trigger must not necessarily be executed from the vehicle but can be resolved by the Master Control (or systems connected to MC). For this use case it would be useful to have the option to send the trigger action from MC to vehicle. The vehicle then confirms it resolved the trigger by setting waitForTrigger Actions exec status to "finished".

**Comment by AntonDueck** [comment]
you are right, you may need a triggerType = MC , an this trigger would be an instant action.

**Comment by MaximilianJohannesObpacher** [comment]
Explanation with scenario and flow diagram:

We have the situation that an AGV wants to go to a node and wait until something is put on top of it. As soon as the load is put on top of the AGV, it should lift its lifter. As both actions happen on the same node the agv needs information when waitForTrigger is finished, so it can raise the lift. The information about loading the AGV being finished comes from a robot. The robot sends the information to the MC. Currently, there is no way of informing the AGV from MC that waitForTrigger should be set to finished.

Therefore, we propose the MC to send a message to the AGV for inform the AGV that the process it was waiting for is finished. The InstantAction for this should be called "trigger". When the message is received bei the AGV it confirms releasing the trigger by setting the waitForTrigger to actionStatus: "FINISHED". The agv can then continue with its pick action. (Diagram related)

The proposed message would look like following:
`
{
"actionId": "trigger-ti18qCti", 
"blockingType": "hard",  
"actionType": "trigger"
"actionParameters": [{"key": "actionId", "value": "waitForTrigger_3-59RRc1dw"}]}
`


![IA_trigger](https://user-images.githubusercontent.com/7281470/134007467-ab2266bf-8e40-4269-a4d8-7dd21a61f0a4.png)

**Comment by philipkirmse** [comment]
Why not working with oder update. Exit edge of node will not get an an release and become part of the base as long the trigger you want to forward to agv is set in MC?

**Comment by collani-bosch** [comment]
@philipkirmse: In my opinion an order update doesn't help, because the order update extends the current order, but doesn't stop the action 'waitForTrigger' to be executed. To use the event 'update' as a trigger would concatenates two different and independent concepts. I think this makes it more complicated.

**Comment by philipkirmse** [comment]
@collani-bosch: I understand your point. But I am not saying use 'waitfortrigger' but using instead release functionality of nodes and edges and order updates. So as we understand VDA5050 description base ends in waiting position until horizon will be released when external trigger meets requirements. So function is a function of traffic management in MC.

So your solution is more the classic way of stopping points and depending of MC architecture it can be easier. In my opinion it would be still possible by VDA5050 because of AGV/project individual actions. So the question is in my opinion still open: Is there a need from MC supplier market that AGV have this function in standard in future or will it stay by VDA5050 a function of MC itself.

**Comment by collani-bosch** [comment]
@philipkirmse: O.K. I got it.

**Comment by philipkirmse** [comment]
@MaximilianJohannesObpacher: I understand now the point which is also a thing for #74. This usual case of multiple actions on one node was not point of view in my argumentation. 

Either it need to be necessary to have a release functionality of individual actions (which is now not possible) or this external trigger signal. Last one even feed in my opinion again another bottleneck e.g. you need to start actions individually on parallel time streams. e.g. you have two conveyor on AGV which waits in a station. Both should start on individual trigger because individual handling.

Suggestion:  So it would be better to have an additional BlockingType (for example called"Externaltrigger") which you can add to each action. A new instantAction which provides an array of in ActionID to AGV can be used as an individual trigger for this blocked Actions. On AGV side this actions in the action array will be only executed by this external trigger.

**Comment by larslue** [comment]
The action "waitForTrigger" should only be used when waiting for a trigger that is issued by the AGV itself.

Otherwise, it wouldn't be clear for the AGV, whether to issue a trigger, or wait for an external trigger.

When the AGV should wait for triggers coming from a 3rd party system to the Master control, the AGV can be oblivious about the fact that it's waiting.

I see it as a shortcoming of the norm that actions at a node cannot drip fed to the robot. As a hack, it might be possible to duplicate the current node of the robot, and add new action to the new node.

**Comment by MaximilianJohannesObpacher** [comment]
@larslue,

you are right regarding the shortcoming. But I think many people have the problem that there are often external triggers. 

The hack doesn't seem good to me, as there would have to be a path with length zero between the points. Seems like that could give some problems as well. Why not have an optional action_parameter with waitForTrigger that is called "allow_remote_resolve" = True and then you can use the trigger action with it.

Otherwise we would need two new actions waitForRemoteTrigger and remoteTrigger. Which does not make sense, as waitForRemoteTrigger and waitForTrigger do exactly the same thing on the AGV, just the resolving condition differs. And what would you do if you have a trigger that can be either triggered remote or on the vehicle? 

From the requirements of us and some other commenters it seems like an extension of waitForTrigger with trigger could be very useful for quite a lot of people. What do you think about it?

**Comment by larslue** [comment]
I agree with you that duplicating the action `waitForTrigger` sounds redundant. As you suggested, the distinction could be done within the `waitForTrigger` action.

It just has to be clear, whether the AGV waits for a remote trigger, or for a local trigger, e.g. hardware button press. So that the AGV can display, whether to ask for a human interaction. Another remote system that is monitoring the fleet could listen to the communication and send an operator to the robot when it hears a `waitForTrigger` action.

Another way of modelling a remote trigger, could be sending the AGV a `pause` action.

**Comment by LukasSchwarz4202** [comment]
We would vote for the originally suggested solution of the _trigger_-instantAction with a parameter _triggerType_:

- It is a simple extension of the interface.
- It is quite useful for simulation purposes. 
- Some supervisors ask for the behavior because in corner cases their routing algorithms rely on this mechanism.

**Comment by KIT-IFL** [comment]
@MaximilianJohannesObpacher, @LukasSchwarz4202 This was discussed in the VDA 5050 core team meeting and the proposal was added to PR https://github.com/VDA5050/VDA5050/pull/365. Feel free to comment there and discuss the topic further.
